src/script/interpreter.cpp

```diff
@@ -1279,18 +1279,28 @@ uint256 GetOutputsHash(const T& txTo)
 } // namespace
 
 template <class T>
-PrecomputedTransactionData::PrecomputedTransactionData(const T& txTo)
+void PrecomputedTransactionData::Init(const T& txTo)
 {
     // Cache is calculated only for transactions with witness
-    if (txTo.HasWitness()) {
+    if (!txTo.HasWitness()) return;
+
+    if (!ready) {
```

Comment:
Supernit: `ready` is faster to check than `HasWitness`, so maybe you want it first (or check `HasWitness`) inside the `!ready` branch.

Reply:
My preferred form after taproot would be:

```
template <class T>
void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut> spent_outputs)
{
    // Save spent outputs for all transactions (not just those with witnesses)
    if (m_spent_outputs.empty()) m_spent_outputs = std::move(spent_outputs);

    // Transactions without witnesses don't need a cache
    if (!txTo.HasWitness()) return;

    // Cache prevouts, sequences and outputs
    if (!ready) {
        m_prevouts_hash = GetPrevoutHash(txTo);
        hashPrevouts = SHA256Uint256(m_prevouts_hash);
        m_sequences_hash = GetSequenceHash(txTo);
        hashSequence = SHA256Uint256(m_sequences_hash);
        m_outputs_hash = GetOutputsHash(txTo);
        hashOutputs = SHA256Uint256(m_outputs_hash);
        ready = true;
    }

    // Cache amounts spent
    if (!m_amounts_spent_ready && !m_spent_outputs.empty()) {
        m_amounts_spent_hash = GetSpentAmountsHash(m_spent_outputs);
        m_amounts_spent_ready = true;
    }
}
```

which is why I modified this commit to early exit after checking for witnesses. Happy to change that back if you think that's better.

Reply:
Heh, I missed this comment, and made this exact change just now in the taproot branch (except a comment).